metatensor-core/src/c_api/io/mod.rs

```diff
@@ -0,0 +1,144 @@
+use std::os::raw::c_void;
+
+mod labels;
+mod tensor;
+
+/// Function pointer to grow in-memory buffers for `mts_tensormap_save_buffer`
+/// and `mts_labels_save_buffer`.
+///
+/// This function takes an existing pointer in `ptr` and a new length in
+/// `new_size`, and grow the allocation. If the pointer is `NULL`, it should
+/// create a new allocation. If it is unable to allocate memory, it should
+/// return a `NULL` pointer. This follows the API of the standard C function
+/// `realloc`, with an additional parameter `user_data` that can be used to hold
+/// custom data.
+#[allow(non_camel_case_types)]
+type mts_realloc_buffer_t = Option<unsafe extern fn(
+    user_data: *mut c_void,
+    ptr: *mut u8,
+    new_size: usize
+) -> *mut u8>;
+
+
+/// Wrapper for an externally managed buffer, that can be grown to fit more data
+struct ExternalBuffer {
+    data: *mut *mut u8,
+    len: usize,
+
+    realloc_user_data: *mut c_void,
+    realloc: unsafe extern fn(*mut c_void, *mut u8, usize) -> *mut u8,
+
+    current: u64,
+}
+
+impl std::io::Write for ExternalBuffer {
+    #[allow(clippy::cast_possible_truncation, clippy::cast_possible_wrap)]
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let mut remaining_space = self.len - self.current as usize;
+
+        if remaining_space < buf.len() {
+            // find the new size to be able to fit all the data
+            let mut new_size = 0;
+            while remaining_space < buf.len() {
+                new_size = if self.len == 0 {
+                    1024
+                } else {
+                    2 * self.len
+                };
+                remaining_space = new_size - self.current as usize;
+            }
+
+            let new_ptr = unsafe {
+                (self.realloc)(self.realloc_user_data, *self.data, new_size)
+            };
+
+            if new_ptr.is_null() {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::OutOfMemory,
+                    "failed to allocate memory with the realloc callback"
+                ));
```

Comment:
Can we always be sure that a null pointer here means out of memory?

Reply:
It's not necessarily out of memory, but it indicates a failure to allocate, which will be an out of memory in 90% of the cases.

This is also the best matching error kind I could find in https://doc.rust-lang.org/std/io/enum.ErrorKind.html

Reply:
Ok, fair